### ISSUE TITLE
analyzer takes many seconds for Snippets+Overrides completions in files that are `part of` "large" libraries

### ISSUE THREAD

**Issue by SecondFlight** [Dart-Code/Dart-Code#4364]
Autocomplete is much faster for me when I use the old analyzer protocol:

https://user-images.githubusercontent.com/6700184/214905760-7bd32cc7-9d06-4d39-a766-c75798ee380e.mp4

This is a proprietary codebase so unfortunately I'm not able to share it. A few things about this codebase which may be relevant:
1. This is a web application which uses dart2js to compile to HTML + JS.
2. This codebase is spread over ~20 packages. There is a main package which imports all the sub-packages.
3. The main library has a file which imports 33 packages, including 8 from the Dart SDK. It has 49 `part` declarations for files within the repository. Most other files don't have imports, and instead use `part of` to inherit the global scope provided by the main file. Most of the other packages in this project do the same.

Please let me know if there is anything else I can share about this codebase.

I've included the info above about imports because the new analyzer protocol seems slowest when trying to autocomplete in the global scope. As you can see in the video, it's much faster once there's a narrower context.

SDK: v2.19 (from Flutter 3.7.0)
OS: Windows 11
Specs: 8th generation i7, 32gb ram

**Comment by DanTup** [comment]
@SecondFlight can you confirm a few things:

- what version of the SDK are you using?
- is this a Flutter or pure Dart project?
- what is your `"dart.autoImportCompletions"` setting set to, and does toggling it to the opposite seem to affect the difference in timings?
- do the items in the completion list seem to be approximately the same, or is there a big difference in the items suggested?
- are you able to reproduce the poor performance (or a significant difference) on a public project (for example opening the root folder of the flutter/futter repo?). For me, if I open flutter/flutter and wait for initial analysis to complete, code completion in `packages/flutter/lib/material.dart` at the very end of the file appears in less than 1 second, so your video does seem very slow (although I'm not using your project and I don't know if there may be something specific triggering this)

Thanks!

**Comment by DanTup** [comment]
If you're able to reproduce it with something shareable that would really help.

Something else that might be useful:

- open your project with it set to LSP (the slow way)
- run the **Dart: Open Analyzer Diagnostics** command
- click **Code Completion** on the left
- back in the editor, invoke the slow code completion
- refresh the diagnostics page and click on the time for the slow completion

This should give you a page like the below, which might at least show which part of completion is taking the time:

![Screenshot 2023-01-26 at 18 38 17](https://user-images.githubusercontent.com/1078012/214921437-855d2a39-5515-4833-bda4-248063e2968a.png)

**Comment by SecondFlight** [comment]
Disabling `dart.enableServerSnippets` fixes the other 5s. It's still noticeably slow compared to the Flutter codebase, but completions happen within 1s.

![image](https://user-images.githubusercontent.com/6700184/214933036-4e62ee2c-1707-42e3-9b91-86775ffc9e87.png)

**Comment by DanTup** [comment]
Oh, that's interesting to know. Are you able to construct a repo you can share? If we're taking 5s to do those snippets, we absolutely need to fix that.

**Comment by SecondFlight** [comment]
I can't do it today, but I'll do it ASAP - probably early next week. I'll ping you when I get some more time to spend on it.

**Comment by DanTup** [comment]
Great, thanks!

**Comment by SecondFlight** [comment]
Hey @DanTup,

I've spent a few hours trying to reproduce the issue, but haven't gotten anywhere yet. I've recreated a similar project structure, but the suggestions in my repro project are still really fast.

Just writing here to say I'm still working on it. Please let me know if you have any suggestions on how I might trace the issue.

**Comment by DanTup** [comment]
@SecondFlight I actually started looking at this today too. I think I understand how snippets can be slow, but I would expect it only to be the very first request (it triggers some expensive work to discover all available files). Can you confirm whether you see it adding on 5s only the first time it's invoked, or all?

As for getting a repro - I don't know how big your project is, but I would probably start by making a copy of it, and then start deleting all the bodies of code (you can use `throw` to avoid having to make functions return the right types) and see whether you can get to something that still reproduces it but might be shareable. I don't know how feasible that is though if your project is massive.

**Comment by SecondFlight** [comment]
@DanTup It does speed up, but it stabilizes to something that's still pretty slow:

![image](https://user-images.githubusercontent.com/6700184/216437568-dc217342-8690-46f8-bd02-14911da55653.png)

Also, I made the mistake of not recording where I had the 10s - I've been seeing 2s-3s recently. `NotImportedContributor` isn't causing issues anymore either which is strange.

Here's the breakdown of that last suggestion. `getSnippets` is the vast majority:
![image](https://user-images.githubusercontent.com/6700184/216438524-7051ea4b-d080-495a-9d22-060681c4392a.png)

**Comment by SecondFlight** [comment]
> As for getting a repro - I don't know how big your project is, but I would probably start by making a copy of it, and then start deleting all the bodies of code (you can use throw to avoid having to make functions return the right types) and see whether you can get to something that still reproduces it but might be shareable. I don't know how feasible that is though if your project is massive.

I appreciate this. I probably still can't share the result, but it's a different angle to attack the problem. Our codebase is ~100k lines of Dart plus HTML and CSS, so it's not small but not unmanageably large either.

**Comment by DanTup** [comment]
Thanks for the info. It would be useful if we could get a better breakdown of timings for the snippets (there are three parts I think are _relatively_ slow, but I wouldn't expect any to take anything like this). Is it possible you could try capturing a CPU profile?

Here's the steps I followed to do this in flutter_gallery:

- Press `F1` to open command palette
- Run the **Preferences: Open Workspace Settings (JSON)** command
- Enable the VM Service for the analyzer by adding something like `"dart.analyzerVmServicePort": 8888,` - this will automatically restart the analysis server
- Wait for initial analysis and invoke completion a few times, to ensure the one we measure is not the first (since for now I'm mostly interested in where it continues to be slow)
- Run the **Dart: Open DevTools** command from the palette
- Paste in `http://localhost:8888/` as the VM Service URL (replacing 8888 with whatever port you used above)
- Click the **CPU Profiler** tab at the top
- Click **Enable Profiler**
- Click the **Record** button
- Invoke completion a _single time_ in a location where it's slow
- Click the **Stop** button
- Click the **Export** button (top right) to save the data

This should produce a JSON file of the profiling data. There shouldn't be anything sensitive in there, but since it's just JSON you can open it up and review it. You may need to zip it up to attach here.

If you want to have a look around, you can click on the CPU Flame Chart and type "snippet" into the search box. `DartSnippetProducer.addImports` is one of the places I thought could be slow, although it was only 4ms for each Flutter snippet in my capture:

![Screenshot 2023-02-03 at 13 21 25](https://user-images.githubusercontent.com/1078012/216613740-4cadad69-7da0-4a1a-80b3-25caaa22326d.png)

Once you're done, be sure to remove the `dart.analyzerVmServicePort` setting.

**Comment by SecondFlight** [comment]
@DanTup Thanks for the instructions. Here's the result:
[dart_devtools_2023-02-03_12_08_29.318.zip](https://github.com/Dart-Code/Dart-Code/files/10581098/dart_devtools_2023-02-03_12_08_29.318.zip)

And here's the timings for good measure:

```
(name: request, elapsed: 0:00:05.070420, elapsedSelf: 0:00:00.000456)
  (name: computeSuggestions, elapsed: 0:00:00.026336, elapsedSelf: 0:00:00.000489)
    (name: DartCompletionManager - ArgListContributor, elapsed: 0:00:00.000820, elapsedSelf: 0:00:00.000820)
    (name: DartCompletionManager - ClosureContributor, elapsed: 0:00:00.000014, elapsedSelf: 0:00:00.000014)
    (name: DartCompletionManager - CombinatorContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - EnumConstantConstructorContributor, elapsed: 0:00:00.000004, elapsedSelf: 0:00:00.000004)
    (name: DartCompletionManager - ExtensionMemberContributor, elapsed: 0:00:00.000173, elapsedSelf: 0:00:00.000173)
    (name: DartCompletionManager - FieldFormalContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - KeywordContributor, elapsed: 0:00:00.000256, elapsedSelf: 0:00:00.000256)
    (name: DartCompletionManager - LabelContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - LibraryMemberContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - LibraryPrefixContributor, elapsed: 0:00:00.000205, elapsedSelf: 0:00:00.000205)
    (name: DartCompletionManager - LocalLibraryContributor, elapsed: 0:00:00.000459, elapsedSelf: 0:00:00.000459)
    (name: DartCompletionManager - LocalReferenceContributor, elapsed: 0:00:00.004680, elapsedSelf: 0:00:00.004680)
    (name: DartCompletionManager - NamedConstructorContributor, elapsed: 0:00:00.000003, elapsedSelf: 0:00:00.000003)
    (name: DartCompletionManager - OverrideContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - RecordLiteralContributor, elapsed: 0:00:00.000007, elapsedSelf: 0:00:00.000007)
    (name: DartCompletionManager - RedirectingContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - StaticMemberContributor, elapsed: 0:00:00.000003, elapsedSelf: 0:00:00.000003)
    (name: DartCompletionManager - SuperFormalContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - TypeMemberContributor, elapsed: 0:00:00.000002, elapsedSelf: 0:00:00.000002)
    (name: DartCompletionManager - UriContributor, elapsed: 0:00:00.000004, elapsedSelf: 0:00:00.000004)
    (name: DartCompletionManager - VariableNameContributor, elapsed: 0:00:00.000001, elapsedSelf: 0:00:00.000001)
    (name: DartCompletionManager - ImportedReferenceContributor, elapsed: 0:00:00.004943, elapsedSelf: 0:00:00.004943)
    (name: DartCompletionManager - NotImportedContributor, elapsed: 0:00:00.014259, elapsedSelf: 0:00:00.014259)
  (name: buildSuggestions, elapsed: 0:00:00.001297, elapsedSelf: 0:00:00.001297)
  (name: mapSuggestions, elapsed: 0:00:00.016931, elapsedSelf: 0:00:00.016931)
  (name: getSnippets, elapsed: 0:00:05.025400, elapsedSelf: 0:00:05.025400)
```

**Comment by DanTup** [comment]
Thanks! Unfortunately this trace is less revealing than I'd hoped. It seems to suggest all the time is spend in analysis and there's not much about snippets. My guess is that computing the snippets is somehow triggering this analysis, however I can't tell why (nor can I reproduce that).

![Screenshot 2023-02-06 at 10 24 03](https://user-images.githubusercontent.com/1078012/216947454-9304cb56-d3f6-4c01-91b3-2b6f9b2aebad.png)

One question I have is whether you're triggering completion my typing characters, or explicitly with Ctrl+Space? If you were doing the former, is it faster if you do the latter? (I'm wondering whether the modification to the file is triggering analysis and the snippets request is just getting stuck behind that, or whether the snippets are really triggering analysis).

My feeling is that you might be right that `part`/`part of` could be contributing to this. By joining all the files together into a single library, any change to any file in that library (such as when you're typing, to trigger code completion) may invalidate far more analysis than if everything was its own library with fewer dependencies. However, if this is only affecting completion/snippets, I suspect there's more to it than just that.

I'll try to finish off the changes I'd started on snippets that may help (they delay computing the "best" imports for snippets instead of doing it up-front), and I may also try to improve the timings being captured to break down snippets a bit more, and we'll see where that gets us.

If you're able to get anything that repros this that can be shared in the meantime, please let me know. Given the trace above, I'm less sure that just deleting all the method bodies will be a reliable way to repro this, as if the time is being spent analyzing code, then removing code may speed it more and somewhat mask the issue.

**Comment by DanTup** [comment]
I've managed to build a repro for this. It has 100 part files and I get around 8s in snippets.

![Screenshot 2023-02-06 at 11 26 20](https://user-images.githubusercontent.com/1078012/216959849-b3a43d98-23ec-405e-8b1b-de550436528a.png)

To answer my last question above, it doesn't seem to matter if there's no modification to the file, just invoking completion in one of the part files takes this long. I'll do some digging, thanks!

**Comment by DanTup** [comment]
@bwilkerson @scheglov interested in your thoughts on this. The issue is this:

- a project has a library with a large number of `parts` (sample project with 100 parts [here](https://github.com/DanTup/dart-code-repro-4364-slow-snippet-completion), readme contains some details on using)
- code completion is invoked inside one of those `part` files
- computing snippets for completion is incredibly slow (~5s for OP, ~8s for me with the sample project above)

The reason this happens is the snippets use `ChangeBuilder` and when modifications are being made to a `part` file, it also creates a builder for the containing library file (to be able to compute `import`s). This triggers analysis of that containing library file which can be slow with a large number of `parts`.

I have an idea for preventing the snippets from trying to compute the `import` edits up-front (by delaying this until the `completion/resolve` request) however on its own, that wouldn't help because this code here would still run while building the main edit:

https://github.com/dart-lang/sdk/blob/91fe18c4e329c14b5c68affb9e87115bfea87cde/pkg/analyzer_plugin/lib/src/utilities/change_builder/change_builder_core.dart#L296-L303

One possible fix for that would be to wrap this code in `if (createEditsForImports)` since we wouldn't need to build edits for imports given the change above. However, I don't know if that's a generally safe thing to do. It would mean there are cases where an edit builder for a part file doesn't have a `libraryEditBuilder` and I'm not sure if that's reasonable (or obvious to callers) that this behaviour changes based on whether imports are being computed.

Another possible fix would be to not use `ChangeBuilder` at all in the snippets, but it would mean losing (or reproducing) some useful behaviour like getting the correct indenting, building linked edits, etc.

(I haven't tested yet, but my suspicion is that this only happens with `part`s and not if all of the code was in-lined, because if it was all in a single file, the resolved unit would be cached because the file is open.. but in this case resolution of the library involves many `part` files that are not open).

**Comment by bwilkerson** [comment]
It's my understanding that we never resolve a single part file in isolation (because we need the import scope in order to resolve the part), but that we instead always resolve a whole library (actually, a library cycle, but that probably isn't relevant to this particular case).

It's my opinion that the issue isn't that we need to access the defining compilation unit of the library in order to create the `ChangeBuilder`, but is that we've lost the reference to the library that we have already resolved, forcing us to repeat that work.

I think the right solution to this problem is for code completion and snippet generation (and possibly other features) to have access to both the `ResolvedLibraryResult` and the `ResolvedUnitResult` in which completion/generation is being performed. Doing so will remove the duplication of effort without requiring more invasive changes.

**Comment by SecondFlight** [comment]
@DanTup Thanks for the additional work on this, and I'm glad you were able to reproduce the issue! I was taking a similar approach to your successful repro, and I'm wondering if I was masking the issue as well... I had 20 libraries with 100 part files each and still wasn't seeing the issue.

I'll stop trying to repro for now as it sounds like you have what you need, but please let me know if I can provide any more information or assistance!

**Comment by DanTup** [comment]
@SecondFlight yep, I think I have enough to make progress for now. I might ask if you're able to test these (using `flutter master` or similar) if I can get some fixes landed to ensure my repro case is similar enough for yours that is solves your problem.

@bwilkerson that makes sense, thanks! I'm working on some changes that have definitely _mostly_ solved this - because each snippet now has the same `ResolvedLibraryResult` available to it (rather than each requesting it's own).

However, it still takes around 1.5s to resolve the library at the start of the request, because the driver cache only keeps hold of open files (and not the parent library). This means the request takes around 1.5s for my sample above, which is better than 1.5s multiplied by the number of snippets that triggered the code, but it's still pretty bad.

I noticed [Code Actions also requires the library](https://github.com/dart-lang/sdk/blob/4e7d71b593971525b0120811107c18b7d25058e6/pkg/analysis_server/lib/src/lsp/handlers/handler_code_actions.dart#L109), so I think it's probably triggering the same thing. When you modify a `part` file, analysis will occur, then VS Code will send a CodeAction request, which will try to get the library (which is not cached) and have to re-resolve it.

I don't know of a good solution to that. It would be nice if `AnalysisDriver.getResolvedLibrary` was able to give us a cached version, or if that would be too much, if `getResolvedUnit`/`getResult` had cached the defining files for each priority file. I don't know what the memory implications of doing that would be though.

**Comment by bwilkerson** [comment]
I think it's the case that it won't require any more memory to produce the `ResolvedLibraryResult` because we've already produced all of the objects anyway. The only memory implication I can think of is that some of the AST structures won't be garbage collectable quite as soon.

**Comment by scheglov** [comment]
I few details about analyzer implementation.

1. We link the whole library cycle, and eventually analyze all libraries in a cycle, but we don't have to resolve all the libraries when analysis of one is requested.
2. We analyze whole libraries, because this is necessary to report `UNUSED_ELEMENT` and likes, private elements can be referenced in parts.
3. It is up to the client to use `getResolvedLibrary()` or `getResolvedUnit()`.

We probably could optimize (2), but IMHO it would be better just to avoid using parts altogether.

**Comment by DanTup** [comment]
>  but we don't have to resolve all the libraries when analysis of one is requested.

I'm not sure I understand why that is, wouldn't everything need to be resolved because each part could be using symbols from each other part?

Anyway, I'm not sure the change I've started making is now the right fix. Lets assume my project is:

```
lib/
    my_library.dart // (contains `part1.dart, part2.dart, ...)
    src/
        part1.dart    // we will invoke code completion in here
        part2.dart
        part3.dart
```

Before my changes, the completion handler would resolve the target file (`part1.dart`). During building of snippets, the DartFileEditBuilder would create a `libraryEditBuilder`, which resolves `my_library.dart` (which is slow and I _think_ resolving all of the `part`s?).

My change was to instead get both the resolved library and the resolved target file up-front at the start of the completion handler, something like:

```dart
final library = requireResolvedLibrary(path);
final unit = library.unitWithPath(path);
```

This allowed me to get the resolved result for `my_library.dart` from `library` when building `libraryEditBuilder` which meant we only resolved the entire library once per completion request, instead of inside each snippet (where it uses the edit builder).

However, while testing I realised that this change means we're now _always_ resolving the library, whereas before it only happened if the Snippet prefixes matched (if you had typed a prefix that matched no snippets, we would never create the edit builders, and never need a resolved library). So while it's faster when there are snippets, it also made every other completion request take 1.5s longer for my sample above.

So, I'm now thinking:

1. If getting the resolved unit for `src/part1.dart` does _not_ resolve the whole library, then we could resolve the library lazily only the first time some snippet needs it, in a way that is reusable by all snippets. This means we'd only pay this cost if it was necessary, and still only once (instead of once per snippet)
2. Do we really need to resolve the whole library (if that's what getting the resolved defining file does) just to be able to produce import edits for the file defining the library?
3. Another change we could make is to produce the imports for snippets "later" (during `"resolve"`) so we don't need them up-front anyway... could we adapt `DartFileEditBuilder` to create the `libraryEditBuilder` lazily (or something) so it just doesn't get triggered here?

**Comment by bwilkerson** [comment]
> ... but we don't have to resolve all the libraries when analysis of one is requested.

What this means is that we don't need to produce resolved ASTs for all of the libraries in a cycle in order to produce resolved ASTs for a single library (where a library is a defining compilation unit and all of the parts of that defining compilation unit). (A cycle is caused by something like 'a.dart' imports 'b.dart', which imports 'c.dart', which then imports 'a.dart', where each dart file is a separate library, each of which might have parts.)

> 1. If getting the resolved unit for src/part1.dart does not resolve the whole library ...

Resolving the unit will resolve the whole library, and then discard everything other than the `ResolvedUnitResult` for the one file being edited.

> 2. Do we really need to resolve the whole library ...

We might be able to avoid that (though I can't think of a way), but we've already resolved all of the files in the library, so there's no reason not to use the information we have.

> 3. Another change we could make is to produce the imports for snippets "later" (during `"resolve"`) ...

That would potentially require us to re-resolve the library on `resolve`, and I don't think it would save us anything.

**Comment by scheglov** [comment]
Note also, that when we do completion, we don't resolve the whole library, and often even not the whole unit. Instead we use `AnalysisDriver.resolveForCompletion()` and try to resolve only the enclosing method body. And we don't compute diagnostics. This allows faster response times.

It is probably detrimental to response times to use heavy-weight resolution of the whole library during completion.

**Comment by DanTup** [comment]
> What this means is that we don't need to produce resolved ASTs for all of the libraries in a cycle in order to produce resolved ASTs for a single library

Ah, got it :)


> That would potentially require us to re-resolve the library on `resolve`, and I don't think it would save us anything.

Resolve only runs when a completion is selected in the UI, so if the user never selects a snippet from the completion list, we'd never run this code. But if we do it up-front while building the completion list, it has to happen for all of the snippets before the user sees any completion. Although, there are some drawbacks - if the user accepts a completion before `resolve` finishes, then its results would be ignored (eg. we wouldn't add the imports). This is already something we just accept for standard auto-import completions.


But, if we have already done all the resolution we need, I agree it'd be better if we could use it. I'm not sure how to achieve this though - I _think_ we'd need to ensure that any time we analyze a part file that is a priority file (in `AnalysisDriver._computeAnalysisResult`) and we cache it's result, we also cache a `ResolvedLibraryResult` so that any handler (like Completion or Code Actions) that need the library could fetch it without re-computing it. Perhaps a cache of "resolved libraries for priority files"?


> Note also, that when we do completion, we don't resolve the whole library, and often even not the whole unit. Instead we use `AnalysisDriver.resolveForCompletion()`

Ah, interesting. We're not using that in LSP right now, we're just calling `getResolvedUnit`/`getResult`. Although I think  with Snippets using the `ChangeBuilder` we wouldn't be able to use this there even without the complication of needing the resolved library (since `DartFileEditBuilderImpl` needs a resolved unit too)?


I'm not sure of the best route to take, but I think some possibilities are:

- find a way to cache libraries for open part files so completion (and CodeActions) handlers can use these without re-computing them
- share a resolved library between snippets, but only compute it once we know there is at least one snippet
- allow `DartFileEditBuilder` to create its `libraryEditBuilder` lazily only if changes need to be made to the library and chance snippets to not produce import edits, moving them to `resolve`
- don't use `ChangeBuilder` at all in the initial Snippets building and use it only to build imports in `resolve` (similar to above)
- ???

I feel like the first option is probably the cleanest, but I don't know what impact on memory it'll have (it'll mean caching more data for the duration of an open `part` file than we do today?) or how accurately/reliably we could invalidate it. I'm interested in your thoughts.

**Comment by bwilkerson** [comment]
We discussed this briefly this morning and one of the ideas we had was to look into whether `ChangeBuilder` could suggest imports based on a parsed AST rather than a resolved AST. The answer might be "no", but it's worth considering.

**Comment by srawlins** [comment]
It sounds like this might be a sizable effort to improve this use case. I think for now we strongly recommend against massive libraries (a library being the defining compilation unit plus all part files). For some definition of "massive" 😄 .

In many cases our "unit of resolution" is the library, and in many cases our "unit of invalidation" (for when code changes and we must throw away a previous resolution result) is also the library. So the IDE tooling will perform much better given 100k lines spread across 100 libraries than 100k lines in a single library (even if spread across parts). This would include completion, navigation, quick fixes, static analysis, etc.

**Comment by DanTup** [comment]
> We discussed this briefly this morning and one of the ideas we had was to look into whether `ChangeBuilder` could suggest imports based on a parsed AST rather than a resolved AST. 

I can have a look and see what's being used and whether it's all available there.

Although even if we did this, it seems like in the case above we'd still _parse_ the library for each snippet kind, so we might still want some way of either caching the library for open files, or being able to share the parsed AST across all snippets (and pass down to their `ChangeBuilder`s?)

(I do have a mostly-working change that would allow a `ChangeBuilder` to be given pre-resolved libraries that speeds this up by only needing to resolve the library once instead of once per snippet, but it's a bit of a weird API change - passing "pre-resolved libraries" down so I don't know if we want to do it if there may be other better ways)


> I think for now we strongly recommend against massive libraries (a library being the concatenation of the defining compilation unit plus all part files). For some definition of "massive" 😄 .

I suspect many users probably don't understand that `part`s are the cause of some of these, and it's not always obvious to us when we get reports of perf issues (especially if the symptom is the server backing up performing this work lots of times, rather than a "slow completion" report). So even if the recommendation is to not structure code this way, I think it's worth us doing something to make sure it doesn't perform badly (even if it's something like a timeout/cancellation for snippets, or even ditching them for part files)? 🤔

**Comment by bwilkerson** [comment]
> ... it seems like in the case above we'd still _parse_ the library for each snippet kind ...

I'd probably approach this using something similar to `CorrectionProducerContext` (`SnippetProducerContext`?) that can hold on to all of the data that snippets need to use. That allows us to cache some of the data for the duration of the request, and makes it easier to add new data later on as the need arises.

> ... it's not always obvious to us when we get reports of perf issues ...

We've talked before about having a set of standard data that users can send to report problems. This would be good information to include in such a mechanism. In the meantime, we could consider adding information to the server diagnostic pages about the number of libraries, use of parts, use of "barrel" files, etc. and ask users to send those stats to us.

> So even if the recommendation is to not structure code this way, I think it's worth us doing something to make sure it doesn't perform badly ...

Absolutely. Sam's comment doesn't mean that we aren't going to do our best to support large libraries. But we know that the time to analyze a library is dependent on its size. We can't really change that reality (though we can try to keep the time linear). And, to the best of our understanding, it probably isn't the use of parts that's a problem, it's the total size of the resulting library that's causing an issue.

The advice is that, in light of that reality, users should keep their libraries to a reasonable size. We don't have numbers yet for what "reasonable" means, but it appears that the OP has a library that our tools aren't currently able to handle well, so restructuring the code base into smaller libraries is likely to be beneficial.

**Comment by DanTup** [comment]
> I'd probably approach this using something similar to `CorrectionProducerContext` (`SnippetProducerContext`?) that can hold on to all of the data that snippets need to use. That allows us to cache some of the data for the duration of the request, and makes it easier to add new data later on as the need arises.

This makes sense, although we need to provide the pre-resolved (/parsed) result into the `ChangeBuilder` which felt a little awkward. I wonder whether it may be useful to have something like a wrapper around the session (`implement`ing session) created for each request, which additionally caches any parsed/resolved unit for it's duration (which is just a single request)? 

I suspect there are probably many places that may accidentally duplicate some work like is happening here, and being able to reuse results without needing specific API changes (eg. just being given some "caching" session) would be worthwhile?


> And, to the best of our understanding, it probably isn't the use of parts that's a problem, it's the total size of the resulting library that's causing an issue.

If I understand correctly, it's not just the size causing an issue here. If the file was a single huge file, the resolved result would be cached (because it's a priority file) and we wouldn't need to fetch the resolved library (because imports go into the same file). Using `parts` means a) we need to additionally get a resolved library to be able to produce edits for imports and b) because there's no caching of libraries we re-resolve it for each snippet producer (because each one creates a `ChangeBuilder` which calls `getResolvedLibrary`).

So while I think it's expected that using `part`s will come with some overhead, I think things are particularly bad here and because the resolution runs for each snippet.

I think I have some things to look into though - 1) whether we can use parsed libraries and 2) reusing the result across snippets (although I'm interested in your thoughts of doing this by providing `ChangeBuilder` a caching-session versus some pre-resolved/parsed results directly).

**Comment by bwilkerson** [comment]
tl;dr I suspect that we need more information before we can make a decision about the best solution, and we might not be able to prioritize that work until after the 3.0 work is farther along.

> I suspect there are probably many places that may accidentally duplicate some work like is happening here, and being able to reuse results without needing specific API changes (eg. just being given some "caching" session) would be worthwhile?

It would be useful to know how many times we use the `AnalysisSession` APIs as part of responding to a request, which would presumably answer the question of whether we're duplicating work.

If we're duplicating work in one place (snippets), then a general mechanism might be overkill. But if it's happening in lots of places then I'd definitely want a general mechanism to solve the problem.

> ... the resolved result would be cached (because it's a priority file) ...

@scheglov 

Seems like we ought to keep all the information we built for the whole library rather than just for a single compilation unit. I understand that we don't resolve the whole library, but we must minimally be resolving the import directives in order to build the name scope for the method containing the completion location, which means that we must have at least parsed the defining compilation unit.

If `ChangeBuilder` _can_ use a parsed compilation unit, then I think we ought to be able to ask the `AnalysisSession` for some kind of semi-resolved library-related data that we can use for this purpose. And if the imports have been resolved, then it might be sufficient for imports even without changes to the `ChangeBuilder`.

**Comment by DanTup** [comment]
@bwilkerson @scheglov FYI, a similar issue came up at https://github.com/Dart-Code/Dart-Code/issues/4421#issuecomment-1479231025. Code completion is very slow, and sometimes it blows the budget resulting in 0 items being returned (and isIncomplete=true). Unfortunately VS Code ignores isIncomplete=true when the results are empty (see https://github.com/microsoft/vscode/issues/155738), so when this occurs no completion items will be shown even if typing continues.

This one is not snippet related, it's overrides. They also use the edit builders (and thus trigger analysis of the containing library in part files). In that particular codebase (Dart-Code-Metrics), that's taking around 2.6s on my machine:

```
(name: OverrideContributor, count: 1, elapsed: 0:00:02.631951, elapsedSelf: 0:00:02.631951)
```

**Comment by bwilkerson** [comment]
We definitely need a solution to this. There's no reason to spend the time to re-resolve everything. We especially need to fix code completion. As a short-term work-around I'd be fine with disabling the override contributor if we're completing in a part file just to ensure that we can provide all of the other completions.

But most of this work (with the possible exception of the completion optimization) needs to wait until after the upcoming deadline.

**Comment by DanTup** [comment]
> As a short-term work-around I'd be fine with disabling the override contributor if we're completing in a part file just to ensure that we can provide all of the other completions.

I can do that if you think it's worthwhile - though I don't know whether it's a good or bad trade-off (users with smaller libraries that aren't so slow might miss these). My message wasn't a prompt to do something, just an FYI in case you hear similar complaints :-)

It could be worth making the budget apply in more places (so we'd never take this long), although there's a part of me that dislikes stopping work early and giving incomplete results (because sometimes I might prefer to wait a little longer and get results than not).

Anyway, something to come back to :-)

**Comment by DanTup** [comment]
This is resolved by @scheglov's change at https://github.com/dart-lang/sdk/commit/92faa1131133e3024c369014f91413f91c8b0fea. In my test case with 100 parts in a single library, the time dropped from 17s (first) and 9s (subsequent) to 760ms (first) and 175ms (subsequent). This was a deliberately extreme case though so the total times are likely much lower for real world code.

The change is part of the Dart SDK (not the VS Code extension), so it will show up after a future SDK update.
